Python callback

Paulchen-Panther opened the pull request:
Also das ist bis jetzt bei mir bei raus gekommen.
Wie gesagt, nicht so viel wie bei dir.
Das ganze sollte dir erst mal verdeutlichen wie ein Callback abläuft.
Hab die service.py vom plugin kodi zum testen genommen

Hier das Beispiel: [service.py](https://gist.github.com/Paulchen-Panther/676cfc368549146ac4a0e2544f98f135)

Wenn Fragen bestehen. Kein scheu. ;-)

Commit:
unneeded

Commit:
Add callback methods

Add registerCallback and unRegisterCallback to the Callback methods.
Remove unnecessary keyword arguments from Wrapper methods

Commit:
destructor changes, add onCompStateChanged

Change the destructor to remove all Callbacks
Add Function onCompStateChanged()
Add the __pluginObj capsule to the dict to get it per PyCapsule_Import

Commit:
callback type definition, inline functions etc.

Add a type definition for the callback enums
Add inline functions to convert string to enum and vice versa
Add helper class acquireGIL. For details see: https://docs.python.org/3/c-api/init.html#non-python-created-threads

Commit:
marco function getPlugin(), unRegisterCallback etc.

Add registerCallback and unRegisterCallback PyCFunction.
Replace Function getPlugin() with smaller Macro
Replace Py_BuildValue("") with Macro Py_RETURN_NONE

Commit:
Connect componentStateChanged to Plugin

brindosch commented:
Ah, super danke :)
Ja das hat doch mal Stil, mit PyCapsule sieht getPlugin() doch gleich viel schlanker aus.  Ich hab leider keinen Weg gesehen wie man eine enum python bekannt machen könnte, so geht es natürlich auch!

Commit:
Moved typedef and inline funcs

Commit:
Fix Compiler Error

Compiling with ENABLED_TEST give this error:
hyperion.h:153:22: error: aggregate ‘std::stringstream ss’ has incomplete type and cannot be defined

Commit:
Moved typedef and inline funcs

Commit:
@brindosch please fix this

I have no idea where to start

Commit:
Fixed

Paulchen-Panther commented:
Ich habe mal die inline functionen in die PluginDefinition.h verschoben. Ist vielleicht ein wenig übersichtlicher.
Das einzigste was du noch anpassen müsstest wäre die TestBlackBorderProcessor.cpp. Die lässt sich mit ENABLED_TESTS nicht compilieren.

brindosch commented:
Alles klar!
danke!

brindosch commented:
Ach, es ist unglaublich. Also erstmal, super!
Allerdings wirft es jetzt segfaults, wenn ich damit sinnvolle Ketten baue.
Also zb, onSettinsChanged: getSettings(). Ich schätze mal das hängt mit dem GIL zusammen? Wenn sich das nicht lösen lässt muss man viele wrapper schreiben. 

Abgesehen davon hab ich was entsetzliches festgestellt, die callbacks freezen den main thread. Das sollte nicht sein. Ich hab jetzt beide Verbindungen auf  Qt:QueuedConnection gesetzt, hilft aber nichts. Dabei sollten die jetzt entkoppelt sein. (Das lässt sich leicht testen in dem man nen time.sleep() in die callback methode hängt.)

Paulchen-Panther commented:
Hallo Brindosch.

Das es segfaults wirft liegt tatsächlich am Gil. Den brauchen wir aber ,ansonsten gibt es Streit zwischen der ganzen Anzahl von callbacks. Ich schau mir das noch mal genauer an. Hatte eigentlich die Hoffnung das du diese Reihenfolge nicht benötigt.

Das mit den Freezes habe ich noch nicht bemerkt. Hab bis jetzt ja nur prints in der Python Ausgabe zum testen benutzt. Werde ich gleich mit einbeziehen in meine Untersuchung.

Das mit den entkoppeln im Callbacks mit Qt:QueuedConnection funktioniert das jetzt eigentlich oder war das ein Problem?

Danke für dein Feedback.

brindosch commented:
>  Hatte eigentlich die Hoffnung das du diese Reihenfolge nicht benötigt

Gäbe es noch andere Tricks wie man das umgehen könnte? Hab überlegt man könnte so wie bei den Effekten ein "setttings" entry im modul beim Start haben welches man einfach aktualisiert und dann den calback called. Wobei... ist das modul nicht global? Die Daten würden dann ja für alle aktiven plugins gelten? 
Es muss irgendwie gehen, Kodi macht das auch. Die nutzen allerdings boost (C Python wrapper + features) ist daher nicht ganz fair der Vergleich.

> Das mit den entkoppeln im Callbacks mit Qt:QueuedConnection funktioniert das jetzt 

Leider hat es nicht geholfen. Es besteht also nocht. Wüsste aber nicht was man noch machen könnte.

Danke

brindosch commented:
Hab den code mal gepusht. Vielleicht siehst du den Fehler. Die Dokumentation sagt aber klar, dass der slot im Ziel thread ausgeführt wird bei einer Queued connection.

Paulchen-Panther commented:
Ich komme auch gerade nicht weiter. Die Idee die ich hatte krieg ich einfach nicht implementiert.
Einfach gesagt, jeden Callback in einen eigenen Thread laufen lassen, damit nicht der Main Thread blockiert wird. Wenn das mal funktionieren sollten, kann man immer noch ein maximum thread counter an callback threads setzten um dann im main thread darauf zu reagieren wenn doch mal ein callback Thread hängt.
Das mit den segfaults im Callback ( onSettinsChanged: getSettings() ) hab ich auch noch nicht durchschaut. Ist auch eine komische Konstellation. Erst ruf Hyperion die Callback Funktion auf, anschließend (mitten drin) dann genau umgekehrt.
Kann nur mit dem GIL zu tun haben. Nur wie tut man in einem aktiven PyObject_CallFunctionObjArgs(), das noch kein rückgabewert besitzt - weil es noch nicht abgelaufen ist - ein erneuten Global Interpreter Lock erzeugen ???

Ich glaube wir denken zu kompliziert. Wir sollten wenn ein Callback ausgeführt wird, nicht wieder zurück laufen (anfordern von main thread) weil wir was vergessen haben, sondern das anfordern was wir wollten.

Für Impressionen deiner seits bin ich jederzeit offen. :-)

brindosch commented:
> Einfach gesagt, jeden Callback in einen eigenen Thread laufen lassen, damit nicht der Main Thread blockiert wird

Nur als Rückfrage, das sollte ja bereits der Fall sein (nicht pro Callback, aber pro Plugin thread). Die slots in der Plugin Klasse sollten jetzt alle vom main thread entkoppelt sein, aber sie blockieren immer noch. Wobei ich nicht sicher bin was genau passiert, wenn man DirectConnection mit QueuedConnection vermischt (Was jetzt der Fall ist, der callback an das websocket geht nicht raus, bevor der callback im Plugin durchgelaufen ist).

Ich hab mal was getestet in welcher Reihenfolge die slots aufgerufen werden. Im folgenden speichere ich Einstellungen (QueuedConnection zwischen main und plugin)

- Send save result to webui now...
- This message is called right after save result emit before method return
- Called from pluginSettingsChanged callback, freezing now for 2s!

Das ist absolut spannend. Im selben Zug mit dem result emit des Speichervorgangs wird der slot in der JsonAPI agearbeitet (Qt::DirectConnection eben - sync - selber thread). Dann geht die emit call methode in die nächste Zeile und erst dann meldet sich das Plugin. Das würde beweisen, dass es eine QueuedConnection ist (Sonst würde sich zuerst das Plugin melden und dann in die nächste Zeile gesprungen)

Hier mit DirectConnection zwischen main thread und plugin (main thread blocking)

- Called from pluginSettingsChanged callback, freezing now for 2s!
- Send save result to webui now...
- This message is called right after save result emit before method return

Wie man sieht, springt er erst in die neue Zeile nachdem auch das callback des plugins returned. Warum zum Henker bleibt der main thread jetzt aber stehen in der QueuedConnection? Die Sache ist doch schon durch!

> Nur wie tut man in einem aktiven PyObject_CallFunctionObjArgs(), das noch kein rückgabewert besitzt 

Schätze mal das geht nicht. Erzeugt der Python -> C Aufruf einen GIL unter der Haube?(!!!) Dann müsste man testen ob der GIL bereits besteht und nutzt ihn einfach mit?! (Dann bräuchte man aber "tieferen" C Python Zugriff - wenn das oben stimmt)

> Wir sollten wenn ein Callback ausgeführt wird, nicht wieder zurück laufen (anfordern von main thread) weil wir was vergessen haben

Alles an Informationen in Argumente packen wäre schon möglich oder dem modul bekant machen (Wobei dem modul das zuzuschreiben welches auf globaler ebene arbeitet und das auf alle laufenden Instanzen Einfluss hätte (großes Fragezeichen) ungeschickt ist?!)

brindosch commented:
Moment, das würde heißen der GIL vom Python Interpreter legt den main thread lahm.

Paulchen-Panther commented:
Die segfaults hab ich beseitigen können.
Mach gleich einen neuen PR.
Die freezes wie gesagt können wir nur beseitigen wenn wir jeden Callback in einen eigenen Thread stecken. Entweder per QThread oder noch besser QtConcurrent. Damit lassen sich jedenfalls funktionen in eigenen Threads stecken.

Paulchen-Panther commented:
> Moment, das würde heißen der GIL vom Python Interpreter legt den main thread lahm.

Das probier ich gleich mal aus. Moment

Paulchen-Panther commented:
> Moment, das würde heißen der GIL vom Python Interpreter legt den main thread lahm.

Nein. Wenn ich im Callback kurz vor dem time.sleep() den GIL wieder freigebe für alle hängt trotzdem der main thread

brindosch commented:
Ja du hast recht, Hab HyperionDaemon in nen thread umgewandelt, es geht trotzdem nichts. (Wenn Python sich den main thread irgendwie krallen würde, gäbe es keine Beeinflussung)

Aber wer erzeugt den thread lock dann? Ich mein das callback an die ui wird rausgegeben. Schafft es aber nicht mehr gesendet zu werden (Da ist genau 1 signal/slot connection dazwischen!)

brindosch commented:
Ok, habs.
Also es ist doch die QueuedConnection. Ich hab den plugin thread einfach mal 3s schlafen gelegt, dadurch dass der slot nicht returned hängt tatschlich der caller thread.

Verstehe es aber nicht. Wenn ein slot in einem Ziel thread ausgeführt wird, kann es dem caller thread doch egal sein wie lange die Aktion dauert.

Paulchen-Panther commented:
[-->Hier<--](https://github.com/brindosch/hyperion.ngBeta/blob/rework/libsrc/plugin/Plugin.cpp#L24) fehlt doch das SIGNAL und SLOT oder irre ich mich da?
Probier mal:
`connect(_prioMuxer, SIGNAL(visiblePriorityChanged( const quint8)), this, SLOT(onVisiblePriorityChanged(const quint8)) );`

brindosch commented:
Leider nein, das ist die qt5 syntax: https://wiki.qt.io/New_Signal_Slot_Syntax/de
Die hat den Vorteil gegenüber der qt4 syntax, dass die Verbindungen während dem Kompilieren überprüft werden anstatt zur Laufzeit Fehler werfen (wie zb xy nicht gefunden)

Paulchen-Panther commented:
![screenshot_20180107_211211](https://user-images.githubusercontent.com/16664240/34653664-88fa578a-f3ef-11e7-9002-46421ec8b561.png)

Irgendwo in den tiefen weiten von hyperion fehlt noch ein qRegisterMetaType() zum Signal Slot. Weiss aber nicht wo. Du hast umgeschrieben. Könntest du da mal schauen?

brindosch commented:
Hast du Image<ColroRgb> in die plugin api gepackt? Ich sehe das nämlich nicht. Da es allerdings so oder so noch kommt - guter Fund :) Habs in die hyperiond gepackt, da es dort ohnehin das erste mal auftritt.

// neuer commit

Zu dem Threads
Ich hab gelesen das subclassing QThread mit slots nicht so toll funktionieren soll. Ist allerdings 10 Jahre die Behauptung und schildert auch nicht ganz das Problem das wir hier haben.

Paulchen-Panther commented:
Heute schaffe ich es leider nicht mehr den PR zumachen. Aber morgen Nachmittag werd ich in hochladen. Bis jetzt funktioniert alles.
Bis morgen
